### DISCUSSION TITLE
Add value change by offset services

### DISCUSSION THREAD

**Discussion by mib1185** [home-assistant/architecture#1057] [Entity Models]
### Current situation

We have many "set value" services like `set_temperature` and `set_humidity` for `climate` or `volume_set` for `media_player` platform.
But in context of an assistant, sentences like "lower volume by 5" or "increase temperature by 2°" would be more common.

### Proposal
Extend all platforms, which actually allows setting a value by a defined service, by corresponding services to change the value by a given offset in a given direction (_increase or decrease_). Separating the offset direction from the offset (_offset is always a positiv value_) is in consideration that intents will most likely be something similar to `<direction> temperature/volume/position by <offset>`.
I've created a POC PR for the climate platform: https://github.com/home-assistant/core/pull/113009

This logic should at least be adopted to the following platform services:
- `climate`: `set_temperature` , `set_humidity` (_see POC PR_)
- `cover`: `set_cover_position`
- `fan`: `set_percentage`
- `humidifier`: `set_humidity`
- `light`: `turn_on` (_brightness only_)
- `media_player`: `volume_set`
- `number`: `set_value`
- `valve`: `set_valve_position`
- `water_heater`: `set_temperature`

---

#### also adopt `time` platform
maybe the `time`platform could also be adopted, but may need some more thoughts, because we need to be unit aware in this case eq. "decrease timer by 10 minutes" or "increase timer by 2 hours"

#### consider possible `min` and `max` values
in case `min` and/or `max` are provided by the entity, these new services may consider them to not set invalid values. but i'm not sure if the should raise a service error or just set to the highest/lowest possible value based on min/max and current value 🤔 

#### allow positiv and negative offset values
these new service accepts positiv and negative offset values (_no need for "offset direction"_) and the intent core needs to consider the wanted number sign based on the requested action (_increase -> positiv, decrease -> negative_)

**Comment by farmio** [comment]
`light.turn_on` already supports `brightness_step` and `brightness_step_pct`, but contrary to your proposal, it works with positive/negative offset values.

**Reply by mib1185** [reply]
i already saw them, but from a native conversation point of view, IMO no one would say "nabu, please change the brightness by -10" 😁 
maybe a possible other solution could by to only have one "change by offset" service, which accepts positiv and negative values and the intent core has to carry about the number sign based on the requested action (_increase -> positiv, decrease -> negative_)

**Reply by tetele** [reply]
> nabu, please change the brightness by -10

That won't be the case. The spoken sentence can undergo some transformations (in the intent code) to prepare the actual values for the service call. So if the service supports an increment (positive or negative), the sentence and intent could, for example, use a `direction` and `offset` pair of slots, with the `offset` being turned to negative when provided to the service call which actually alters the value if the `direction` is `"decrease"`.

**Comment by TheFes** [comment]
I've been using Google Home devices with voice control since they have been introduced now, I both for volume or brightness I have never thought of using the step instead of the target
I've always used `set player volume to 25%` or `set light to 50%`.

So I'm not opposed having step being supported, but the current versions should be supported as well.

**Comment by tetele** [comment]
I like this idea very much (i.e. the generalization of these increment/decrement services).

**Comment by frenck** [comment]
This has been discussed quite a few times before, and recently, effort has been made to implement it for cover positions (for example).

We've stopped pursuing this, as it causes lots of issues (including current implementations). (@emontnemery can provide more context on this one). And even for recent attempt to implement this for media player volumes, caused quite some challenges (although we have a path forward).

Besides technical challenges, it also imposes problems with async handling integrations, like polling integrations, that do not natively support incremental changes (which are the majority). As in those cases, we have to assume the current state reflects the actual state when we increment/decrement for them.

There is nothing stopping all of this from being done by consumers with the existing logic available (e.g., using templates). At that point, it becomes a user problem, which might be just good enough for that specific usecase.


../Frenck
